### PR TITLE
Class locations

### PR DESCRIPTION
**Pull request by AurumTheEnd**

### CODE REVIEW

**Comment by AurumTheEnd** [comment]
@callumhutchy Hi, I've noticed your fork and changeset when working on the repo, and it's quite an interesting one. Would you be interested in polishing the changes and somehow integrating them into JRD? Or was this just a one-off endeavor?

**Comment by callumhutchy** [comment]
It's been a while as you can see, but I'll have a look this evening and try and tidy some stuff up and create a PR

**Comment by AurumTheEnd** [comment]
Alright, thanks for the swift reply. As a heads-up, some of the required polish will have to include:
 - Remove Eclipse-related project files and add their file patterns to `.gitignore`
 - Remove the output `.csv` file (although it was good to have it to see what your changeset did)
 - Extract the purely cosmetic code changes into a separate commit

Thanks :)

**Comment by callumhutchy** [comment]
I spent some time tonight updating my changes to the latest master, I realise that the code that is on my fork might not be the latest and I'll check tomorrow on my work machine if there are some worth additions.

Currently, it only writes to a file on the initial load, I was writing on the reload class button as well in another version and saving files with timestamps in the filename so as not to overwrite classes that only appeared in a previous version.

For my purposes I wanted to keep track of all classes I came across and so refreshing allowed me to catch new classes as they were loaded.

I could easily increase the scope but for a first contribution probably best to keep it short and sweet.

**Comment by judovana** [comment]
Hello! Thanx a lot for an effort. it is highly appreciated!
AurumTheEnd aka Prokopl Tunel (try that in google transalte from CZ :) aka Ondra, will review.
I have few nits:
 * there is still couple of files to ignore and remove:
   * .project, decompiler_agent/.classpath, decompiler_agent/.project , images/.project, runtime-decompiler/.classpath,  runtime-decompiler/.project
 *  there are some code formatting chnages which should not be done - mainly the long line wraps. (but  many are good)
    *  We currently do not have code format file pushed in nor maven code format plugin included. So there is no press. Ondra or Soti  (ilbinek) will add it and then whole project will be reformatted.
  *  images/pom.xml  the chgange should be reverted
  *  maybe use different delimiter shoudl be used for sending from agent - the "," may be part of path. Maybe ":" or ";" or "|"?
  *   private List<String[]> loadedClassInfos; I think the String[] deserve to be replaced by proper wrapper class with fields instead of items in String array.  Then also the parseClassInfos can be placed properly here. We can do it later if you do not wish to bother.
  *  maybe the api can be later enhanced to get info of only one class
  
I think the writing to the file can be kept for now. Once it is in you, or me or Ondra or Soti will change it to print to the class text view in gui mode, and to stdout in CLI mode (so you will be able to stream/save to file anyway). Also the support for to "FileSystem VM" can be added by us.

Thanx a lot for idea again. TBH, I did nto know it is possible to write out what file the class was at the end read from.  And it was always missing piece of information

**Comment by callumhutchy** [comment]
All good stuff, I'll make the amendments this evening.

I reviewed the uncommited code on my work machine and it was a hacky mess so best to leave it out for now. What it does include is the name of the class loader that was used to load each class, which was another particular bit of info I was looking for. If that is of any use we can possibly include it at a later date.

File: runtime-decompiler/src/main/java/org/jrd/backend/core/DecompilerRequestReceiver.java [hunk 1]
```diff
@@ -199,7 +202,25 @@ private String getAllLoadedClassesAction(String hostname, int listenPort, String
...
+            List<String[]> arrayOfClasses = parseClassInfos(reply.response);
+
+            VmDecompilerStatus status = new VmDecompilerStatus();
+            status.setHostname(hostname);
+            status.setListenPort(reply.port);
+            status.setVmId(vmId);
+            status.setLoadedClassInfos(arrayOfClasses);
+            vmManager.getVmInfoByID(vmId).replaceVmDecompilerStatus(status);
+        } catch (Exception ex) {
+            OutputController.getLogger().log("Jaboody");
```
Comment (AurumTheEnd) [inline runtime-decompiler/src/main/java/org/jrd/backend/core/DecompilerRequestReceiver.java:219]: Jaboody? :D

**Review by AurumTheEnd — COMMENTED** [review]
Food for thought for your refactoring. Note that because I created the PR, I cannot request changes. Thanks, GitHub.

File: runtime-decompiler/src/main/java/org/jrd/backend/core/DecompilerRequestReceiver.java (hunk 1 above)
Comment (callumhutchy) [reply runtime-decompiler/src/main/java/org/jrd/backend/core/DecompilerRequestReceiver.java:219]: Aha some classic console log debugging, must have been watching some jaboody dubs while writing 😅

**Comment by callumhutchy** [comment]
> Food for thought for your refactoring. Note that because I created the PR, I cannot request changes. Thanks, GitHub.

All comments are fair, I'll correct them later

**Comment by AurumTheEnd** [comment]
Also please be aware that I have not been able to get the feature to work in the current state, and because of the rebase I can't really checkout the original May commit. Might have something to do with the plethora of changes that JRD went through during the summer since May.

**Comment by AurumTheEnd** [comment]
A vision we have discussed with other collaborators regarding this feature today was that of including your functionality of getting additional class information within the current CLASS_NAMES/CLASSES request for the agent, rather than creating a second request type that gets queried after the names.

Obviously, this would entail changing the return type from String to the aforementioned new wrapper pojo class all the way through, but would pave the way for an easy integration of this feature into the GUI and CLI.

Only counter-point mentioned was a possibility of further slowing down the action of class loading (which is already why it's on a separate thread), but from my debugging it looks like getters shouldn't be of any issue in this regard.

**Comment by AurumTheEnd** [comment]
Decided to implement the suggested changes to a working condition myself in #119, because you @callumhutchy made us aware of additional information available in the JVM.

You can either choose to implement the changes here, or close this pull request and leave the feature-set in our hands. You will be credited either way.

**Comment by callumhutchy** [comment]
> Decided to implement the suggested changes to a working condition myself in #119, because you @callumhutchy made us aware of additional information available in the JVM.
> 
> You can either choose to implement the changes here, or close this pull request and leave the feature-set in our hands. You will be credited either way.

Your version is much cleaner, so best to go with that.

The class loader info is also quite handy for people to know in case it's not the default one, but that's a separate feature

**Comment by AurumTheEnd** [comment]
> The class loader info is also quite handy for people to know in case it's not the default one, but that's a separate feature.

It can simply be another commit in my PR. Would you @callumhutchy mind sharing what information about the classloader you were after, and the sequence of getters that lead you to it?

**Comment by AurumTheEnd** [comment]
Surpassed by in-house #119.

**Pull request closed without merging**